api/sso.go

```diff
@@ -0,0 +1,407 @@
+			}
+		} else {
+			return badRequestError("A SSO provider with this resource ID (%s) already exists", params.ResourceID)
+		}
+	}
+
+	existingProvider, err := models.FindSAMLProviderByEntityID(a.db, metadata.EntityID)
+	if err != nil && !models.IsNotFoundError(err) {
+		return err
+	}
+	if existingProvider != nil {
+		return badRequestError("SAML Identity Provider with this EntityID (%s) already exists", metadata.EntityID)
+	}
+
+	provider := &models.SSOProvider{
+		// TODO handle Name, Description, Attribute Mapping
+		SAMLProvider: models.SAMLProvider{
+			EntityID:    metadata.EntityID,
+			MetadataXML: string(rawMetadata),
+		},
+	}
+
+	if params.ResourceID != "" {
+		provider.ResourceID = &params.ResourceID
+	}
+
+	if params.MetadataURL != "" {
+		provider.SAMLProvider.MetadataURL = &params.MetadataURL
+	}
+
+	for _, domain := range params.Domains {
+
+		existingProvider, err := models.FindSSOProviderByDomain(a.db, domain)
+		if err != nil && !models.IsNotFoundError(err) {
+			return err
+		}
+		if existingProvider != nil {
+			return badRequestError("SSO Domain '%s' is already assigned to an SSO identity provider (%s)", domain, existingProvider.ID.String())
+		}
+
+		provider.SSODomains = append(provider.SSODomains, models.SSODomain{
+			Domain: domain,
+		})
+	}
+
+	if err := a.db.Transaction(func(tx *storage.Connection) error {
+		return tx.Eager().Create(provider)
+	}); err != nil {
+		return err
+	}
+
+	return sendJSON(w, http.StatusCreated, provider)
+}
+
+// adminSSOProvidersGet returns an existing SAML Identity Provider in the system.
+func (a *API) adminSSOProvidersGet(w http.ResponseWriter, r *http.Request) error {
+	provider := getSSOProvider(r.Context())
+
+	return sendJSON(w, http.StatusOK, provider)
+}
+
+// adminSSOProvidersUpdate updates a provider with the provided diff values.
+func (a *API) adminSSOProvidersUpdate(w http.ResponseWriter, r *http.Request) error {
+	ctx := r.Context()
+
+	body, err := getBodyBytes(r)
+	if err != nil {
+		return internalServerError("Unable to read request body").WithInternalError(err)
+	}
+
+	var params CreateSSOProviderParams
+	if err := json.Unmarshal(body, &params); err != nil {
+		return badRequestError("Unable to parse JSON").WithInternalError(err)
+	}
+
+	if err := params.validate(true /* <- forUpdate */); err != nil {
+		return err
+	}
+
+	dirty := false
+
+	provider := getSSOProvider(ctx)
+
+	if params.ResourceID != "" && (provider.ResourceID == nil || params.ResourceID != *provider.ResourceID) {
+		// resource ID is being updated
+
+		existingProvider, err := models.FindSSOProviderByResourceID(a.db, params.ResourceID)
+		if err != nil && !models.IsNotFoundError(err) {
+			return err
+		}
+
+		if existingProvider != nil {
+			return badRequestError("SSO Provider Resource ID (%s) can't be updated, already assigned to another provider (%s)", params.ResourceID, existingProvider.ID.String())
+		}
+
+		provider.ResourceID = &params.ResourceID
+		dirty = dirty || true
```

Comment:
Whenever the passed parameters don't match the state in the DB.

Reply:
Oh ok. Might be missing something here but is there any reason why we don't directly assign `dirty = true`  then? Seems like it'd always be false since `true || false` is always `true`

Reply:
Yeah I'll rewrite it, not sure why I did this. :D

Reply:
Changed it to `modified` and also removed the `|| true` code.